File: ProjectCode/front/lib/api/route_api.dart
// ========================================================================
// 배리어프리 보행자 네비게이션 — API 서비스
// ========================================================================
// route_api.dart — 백엔드 경로 탐색 API 통신 담당
// ========================================================================

import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:latlong2/latlong.dart';

import '../models/route_response.dart';

/// ── API 통신 서비스 ──
///
/// 백엔드 Django 서버와의 HTTP 통신을 담당한다.
/// 싱글턴 패턴은 사용하지 않고, 테스트 용이성을 위해
/// baseUrl 과 http.Client 를 외부에서 주입할 수 있도록 설계했다.
class RouteApiService {
  /// 백엔드 서버 기본 주소
  /// ★ 실제 배포 시 이 값을 환경 변수나 설정 파일에서 읽도록 변경
  final String baseUrl;

  /// HTTP 클라이언트 (테스트 시 Mock 주입 가능)
  final http.Client _client;

  /// 요청 타임아웃 (초)
  static const int _timeoutSeconds = 15;

  RouteApiService({
    this.baseUrl = 'http://localhost:8000/api/v1',
    http.Client? client,
  }) : _client = client ?? http.Client();

  // ═══════════════════════════════════════════════
  // 경로 탐색 API 호출
  // ═══════════════════════════════════════════════

  /// 출발지/도착지 좌표로 배리어프리 최단 경로를 요청한다.
  ///
  /// [start] 출발지 좌표 (LatLng)
  /// [end]   도착지 좌표 (LatLng)
  ///
  /// 반환: [RouteResponse] — 성공 시 경로 GeoJSON + 총 거리 포함
  ///
  /// ★ POST /api/v1/route/ 엔드포인트를 호출한다.
  Future<RouteResponse> fetchRoute({
    required LatLng start,
    required LatLng end,
  }) async {
    final url = Uri.parse('$baseUrl/route/');

    // ── 요청 바디 구성 ──
    final body = jsonEncode({
      'start_lat': start.latitude,
      'start_lng': start.longitude,
      'end_lat': end.latitude,
      'end_lng': end.longitude,
    });

    try {
      // ── POST 요청 전송 ──
      final response = await _client
          .post(
            url,
            headers: {
              'Content-Type': 'application/json',
              'Accept': 'application/json',
            },
            body: body,
          )
          .timeout(Duration(seconds: _timeoutSeconds));

      // ── 응답 파싱 ──
      final jsonData = jsonDecode(utf8.decode(response.bodyBytes))
          as Map<String, dynamic>;

      if (response.statusCode == 200) {
        return RouteResponse.fromJson(jsonData);
      }

      // 4xx / 5xx 에러 처리
      return RouteResponse(
        success: false,
        error: jsonData['error'] as String? ??
            '서버 오류가 발생했습니다. (${response.statusCode})',
      );
    } on SocketException {
      return const RouteResponse(
        success: false,
        error: '서버에 연결할 수 없습니다.\n네트워크 연결을 확인하세요.',
      );
    } on http.ClientException catch (e) {
      return RouteResponse(
        success: false,
        error: '네트워크 오류: ${e.message}',
      );
    } on FormatException {
      return const RouteResponse(
        success: false,
        error: '서버 응답을 파싱할 수 없습니다.',
      );
    } catch (e) {
      return RouteResponse(
        success: false,
        error: '알 수 없는 오류: $e',
      );
    }
  }

  // ═══════════════════════════════════════════════
  // 헬스체크
  // ═══════════════════════════════════════════════

  /// 백엔드 서버 상태를 확인한다.
  Future<bool> healthCheck() async {
    try {
      final url = Uri.parse('$baseUrl/health/');
      final response = await _client
          .get(url)
          .timeout(const Duration(seconds: 5));
      return response.statusCode == 200;
    } catch (_) {
      return false;
    }
  }

  /// 리소스 해제
  void dispose() {
    _client.close();
  }
}

File: ProjectCode/front_react/src/api/routeApi.ts
// ========================================================================
// 배리어프리 보행자 네비게이션 — API 서비스
// ========================================================================
// routeApi.ts — 백엔드 경로 탐색 API 통신 담당
// ========================================================================

import type { LatLngTuple } from "leaflet";

import type { RouteResponse } from "../models/routeResponse";
import { parseRouteResponse } from "../models/routeResponse";

// ── 백엔드 서버 기본 주소 ──
//   개발: vite.config.ts 의 proxy 설정으로 /api/* 가 Django 로 전달됨
//   운영: VITE_API_BASE_URL 환경 변수로 절대 주소 지정 가능
const DEFAULT_BASE_URL =
  import.meta.env.VITE_API_BASE_URL ?? "/api/v1";

// ── 요청 타임아웃 (밀리초) ──
const TIMEOUT_MS = 15_000;


// ========================================================================
// 경로 탐색 API 호출
// ========================================================================

export interface FetchRouteArgs {
  start: LatLngTuple;  // [위도, 경도]
  end: LatLngTuple;
  baseUrl?: string;
  signal?: AbortSignal;
}

/**
 * 출발지/도착지 좌표로 배리어프리 최단 경로를 요청한다.
 *
 * ★ POST /api/v1/route/ 엔드포인트를 호출한다.
 */
export async function fetchRoute({
  start,
  end,
  baseUrl = DEFAULT_BASE_URL,
  signal,
}: FetchRouteArgs): Promise<RouteResponse> {
  const url = `${baseUrl}/route/`;

  // ── 요청 바디 구성 ──
  const body = JSON.stringify({
    start_lat: start[0],
    start_lng: start[1],
    end_lat: end[0],
    end_lng: end[1],
  });

  // ── 타임아웃 처리 (외부 signal 과 결합) ──
  const timeoutController = new AbortController();
  const timeoutId = setTimeout(() => timeoutController.abort(), TIMEOUT_MS);

  // 외부 signal 이 abort 되면 내부 controller 도 abort
  if (signal) {
    if (signal.aborted) timeoutController.abort();
    else signal.addEventListener("abort", () => timeoutController.abort());
  }

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Accept": "application/json",
      },
      body,
      signal: timeoutController.signal,
    });

    // ── 응답 파싱 ──
    let json: unknown;
    try {
      json = await response.json();
    } catch {
      return {
        success: false,
        totalDistance: 0,
        totalDistanceDisplay: "",
        nodeCount: 0,
        pathNodes: [],
        routeGeoJson: null,
        error: "서버 응답을 파싱할 수 없습니다.",
      };
    }

    if (response.ok) {
      return parseRouteResponse(json);
    }

    // ── 4xx / 5xx 에러 처리 ──
    const errorMsg =
      (json as { error?: string })?.error ??
      `서버 오류가 발생했습니다. (${response.status})`;
    return {
      success: false,
      totalDistance: 0,
      totalDistanceDisplay: "",
      nodeCount: 0,
      pathNodes: [],
      routeGeoJson: null,
      error: errorMsg,
    };
  } catch (err) {
    // ── 네트워크 / 타임아웃 / Abort 등 ──
    let message = "알 수 없는 오류가 발생했습니다.";
    if (err instanceof DOMException && err.name === "AbortError") {
      message = "요청이 시간 초과되었거나 취소되었습니다.";
    } else if (err instanceof TypeError) {
      // fetch 실패는 보통 네트워크 단절
      message = "서버에 연결할 수 없습니다.\n네트워크 연결을 확인하세요.";
    } else if (err instanceof Error) {
      message = `네트워크 오류: ${err.message}`;
    }
    return {
      success: false,
      totalDistance: 0,
      totalDistanceDisplay: "",
      nodeCount: 0,
      pathNodes: [],
      routeGeoJson: null,
      error: message,
    };
  } finally {
    clearTimeout(timeoutId);
  }
}


// ========================================================================
// 헬스체크
// ========================================================================

export async function healthCheck(
  baseUrl: string = DEFAULT_BASE_URL
): Promise<boolean> {
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 5_000);
    const response = await fetch(`${baseUrl}/health/`, {
      signal: controller.signal,
    });
    clearTimeout(timeout);
    return response.ok;
  } catch {
    return false;
  }
}
